lib/rack/session/cookie.rb

```diff
@@ -253,6 +259,14 @@ def persistent_session_id!(data, sid = nil)
         data
       end
 
+      def set_cookie(request, response, cookie)
+        return if request.session.to_h == request.get_header(RACK_SESSION_WAS) &&
+                  !request.get_header(RACK_SESSION_OPTIONS).fetch(:renew, false) &&
+                  !cookie[:expires]
```

Comment:
I would reverse the order of operations here. Only do the `request.session` conversion if the other two cheaper checks pass.

As mentioned in the other comment, instead of `request.session.to_h`, I would compare using marshal/json to avoid issues when modifying nested structures.